data/plugins.ts

```diff
@@ -8,14 +33,77 @@ async function fetchPluginData(): Promise<string[]> {
 	for (let index = 0; index < mirrors.length; index++) {
 		try {
 			plugins = await (await fetch(mirrors[index])).json()
-			return Object.values(plugins)
+			return plugins
 		} catch (error) {
 			continue
 		}
 	}
-	throw new Error("Cannot find avalible Github mirror");
+	throw new Error('Cannot find avalible Github mirror')
+}
+
+function convertToDomId(id: string): string {
+	return id.toLowerCase().replace(/[.]/g, '-')
+}
+
+async function getTrendingPlugins(
+	plugins: Manifest,
+	topn: number
+): Promise<JoplinPlugin[]> {
+	const result = []
+	for (const pluginId in plugins) {
+		result.push({
+			id: pluginId,
+			downloadCount: plugins[pluginId].downloadCount,
+		})
+	}
+
+	return result
+		.sort((a, b) => b.downloadCount - a.downloadCount)
+		.slice(0, topn)
+		.map((item) => {
+			return {
+				domId: convertToDomId(item.id),
+				...plugins[item.id],
+			}
+		})
 }
 
-export default async function getPlugins(): Promise<string[]> {
-	return await fetchPluginData();
-};
+async function getPluginData(): Promise<Manifest> {
+	const period = 'last-week'
+
+	const rawPlugins = await fetchPluginData()
+	for (const pluginId in rawPlugins) {
+		const packageName: string = rawPlugins[pluginId]._npm_package_name
+		const downloadStat: { downloads: number } = await (
+			await fetch(
+				`https://api.npmjs.org/downloads/point/${period}/${packageName}`
```

Comment:
Does this include the number of times users downloaded the package (does the Joplin plugin repository download from NPM then distribute copies of the downloaded package, or are users getting plugins directly from NPM)?

Reply:
Looking at the plugins repository, it looks like plugins are hosted on GitHub, but that GitHub auto-fetches from NPM.
https://github.com/joplin/plugins/blob/master/plugins/io.github.personalizedrefrigerator.joplin-vimrc/plugin.jpl

I see very different numbers in `stats.json` (on GitHub), [see link](https://github.com/joplin/plugins/blob/da04b30141ed309012141a19a331ca421fca41b4/stats.json#L2638), than on NPM...

Reply:
users are getting plugins from the @joplin/plugin repo, it also has a stats in that repo. I'm directly fetching from npm instead of stats is because that the update of stats is a little bit slow, and it can cause unwanted errors when the plugin list is updated but the stats is not. 
If this is different then I can switch back to stats.

Reply:
NPM lists my joplin-plugin-custom-codemirror-vimrc plugin as having [85 downloads](https://www.npmjs.com/package/joplin-plugin-custom-codemirror-vimrc) while [stats.json says it has only 4 downloads.](https://github.com/joplin/plugins/blob/da04b30141ed309012141a19a331ca421fca41b4/stats.json#L2638). I suggest switching back to stats.

Reply:
The reason why it differs so much is probably the stats.json is counting all the downloads in that version, and I'm fetching the download count for the last week. If we switch back to stats, the trending section will be broken because we cannot get the accurate number of downloads within a certain period.

Reply:
I don't think 85 downloads makes sense for my `vimrc` plugin that I released just a few days ago. The 5 **cumulative** downloads listed in `stats.json` make much more sense.

Reply:
I have replaced the downloadCount from npm to the stats, now the trending is determined by downloadCount / (time.now() - releaseTime). I think keeping a copy of the stats.json from the previous week would help, and I'm working on it.